colorbleed/plugins/maya/publish/validate_animation_out_set_related_node_ids.py

```diff
@@ -18,72 +17,116 @@ class ValidateAnimationOutSetRelatedNodeIds(pyblish.api.InstancePlugin):
         for node in nodes:
+            node_type = cmds.nodeType(node)
             node_id = lib.get_id(node)
+
+            if node_type == "mesh" and not node_id:
+                invalid.append(node)
+                continue
+
             if not node_id:
-                invalid_items.append(node)
+                continue
+
+            root_id = cls.get_history_root_id(node=node)
+            if root_id is not None:
+                asset_id = cls.to_item(node_id)
+                if root_id != asset_id:
+                    invalid.append(node)
+
+        return invalid
+
+    @classmethod
+    def get_history_root_id(cls, node):
+        """
+
+        Get the original node ID when a node has been deformed
+        Args:
+            node (str): node to retrieve the
+
+        Returns:
+            str: the asset ID as found in the database
+        """
+
+        asset_id = None
+        node = cmds.ls(node, long=True)[0]
+
+        # We only check when the node is *not* referenced
+        if cmds.referenceQuery(node, isNodeReferenced=True):
+            return
+
+        # Find all similar nodes in history
+        history = cmds.listHistory(node)
+        node_type = cmds.nodeType(node)
+        similar_nodes = cmds.ls(history, exactType=node_type, long=True)
 
-        # TODO: Should we check whether the ids are related to the rig's asset?
+        # Exclude itself
+        similar_nodes = [x for x in similar_nodes if x != node]
 
-        # Log invalid item ids
-        if invalid_items:
-            for item_id in sorted(invalid_items):
-                cls.log.warning("Found invalid item id: {0}".format(item_id))
+        # The node *must be* under the same parent
+        parent = cls.get_parent(node)
+        similar_nodes = [i for i in similar_nodes if
+                         cls.get_parent(i) == parent]
 
-        return invalid_items
+        # Check all of the remaining similar nodes and take the first one
+        # with an id and assume it's the original.
+        for similar_node in similar_nodes:
+
+            _id = lib.get_id(similar_node)
+            if not _id:
+                continue
+
+            asset_id = cls.to_item(_id)
+            break
+
+        return asset_id
+
+    @classmethod
+    def repair(cls, instance):
+
+        for node in cls.get_invalid(instance):
+            # Get node ID and the asset ID part
+            node_id = lib.get_id(node)
+            asset_id = cls.to_item(node_id)
+
+            # Get root asset ID
+            root_id = cls.get_history_root_id(node=node)
+
+            # Replace errored ID with good ID
+            new_id = node_id.replace(asset_id, root_id)
+
+            cmds.setAttr("%s.cbId" % node, new_id, type="string")
 
     @staticmethod
-    def to_item(id):
+    def to_item(_id):
         """Split the item id part from a node id"""
-        return id.rsplit(":", 1)[0]
+        return _id.split(":", 1)[0]
+
+    @staticmethod
+    def get_parent(node):
```

Comment:
Let's not make this a `staticmethod` but just make it a separate function at the top to simplify things.